### PR TITLE
Implement noneOrFail #4841

### PR DESCRIPTION
**Pull request by tyrcho**

### CODE REVIEW

**Comment by tyrcho** [comment]
makes sense, I'll try that tomorrow

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -3649,6 +3649,14 @@ object ZIO extends ZIOCompanionPlatformSpecific {
    */
   val none: UIO[Option[Nothing]] = succeedNow(None)
 
+  /**
+   * Lifts an Option into a IO.
+   * If the option is empty it succeeds with Unit.
+   * If the option is defined it fails with an error adapted with f.
+   */
+  def noneOrFail[E, O](o: Option[O], f: O => E): IO[E, Unit] =
```
[the remark this answers is not in the document]
Comment (tyrcho) [reply core/shared/src/main/scala/zio/ZIO.scala]: I updated the code. Now the 2 methods are available.

**Comment by tyrcho** [comment]
hey @adamgfraser , anything still needed before the merge ?

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 2]
```diff
@@ -3649,6 +3649,22 @@ object ZIO extends ZIOCompanionPlatformSpecific {
...
+   */
+  def noneOrFail[E](o: Option[E]): IO[E, Unit] =
+    getOrFailUnit(o).flip
+
+  /**
+   * Lifts an Option into a IO.
+   * If the option is empty it succeeds with Unit.
+   * If the option is defined it fails with an error adapted with f.
+   */
+  def noneOrFailWith[E, O](o: Option[O], f: O => E): IO[E, Unit] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZIO.scala:3665]: I would curry the arguments here for improved type inference.

File: core/shared/src/main/scala/zio/IO.scala [hunk 3]
```diff
@@ -730,6 +730,18 @@ object IO {
...
+  /**
+   * @see See [[zio.ZIO.noneOrFail]]
+   */
+  def noneOrFail[E](o: Option[E]): IO[E, Unit] =
+    ZIO.noneOrFail(o)
+
+  /**
+   * @see See [[zio.ZIO.noneOrFailWith]]
+   */
+  def noneOrFailWith[E, O](o: Option[O], f: O => E): IO[E, Unit] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/IO.scala:742]: You could have variants for some of the other type aliases that specialized the `E` type to `Throwable`.

**Review by adamgfraser — CHANGES_REQUESTED** [review]
Looks great! A couple of minor comments. Could you also pull in the latest upstream changes? We've had some changes to our CI infrastructure recently so want to make sure everything is running. Thanks again for your work on this!

**Comment by tyrcho** [comment]
Thanks for the review !

I'll have a look at the changes you requested, I think they all make sense. I don't know when I can handle these, this week is going to be quite busy.

**Comment by tyrcho** [comment]
I rebased from master branch, currified, and added an alias in RIO and Task. The Task alias really makes sense !

Anything else @adamgfraser ?

**Review by adamgfraser — APPROVED** [review]
Looks great!

**Pull request merged**